graphql_compiler/schema_transformation/make_query_plan.py

```diff
@@ -0,0 +1,274 @@
+        )
+        if new_selection is not selection:  # Changes made somewhere down the line
+            if not made_changes:
+                made_changes = True
+            else:
+                # Change has already been made, but there is a new change. Implies that multiple
+                # fields have the @output directive with the desired name
+                raise GraphQLValidationError(
+                    u'There are multiple @output directives with the out_name "{}"'.format(
+                        field_out_name
+                    )
+                )
+        new_selections.append(new_selection)
+
+    if made_changes:
+        new_ast = copy(ast)
+        new_ast.selection_set = ast_types.SelectionSet(selections=new_selections)
+        return new_ast
+    else:
+        return ast
+
+
+def _is_output_directive_with_name(directive, out_name):
+    """Return whether or not the input is an @output directive with the desired out_name."""
+    if not isinstance(directive, ast_types.Directive):
+        raise AssertionError(u'Input "{}" is not a directive.'.format(directive))
+    return (
+        directive.name.value == OutputDirective.name and
+        directive.arguments[0].value.value == out_name
+    )
+
+
+def _get_in_collection_filter_directive(input_filter_name):
+    """Create a @filter directive with in_collecion operation and the desired variable name."""
+    return ast_types.Directive(
+        name=ast_types.Name(value=FilterDirective.name),
+        arguments=[
+            ast_types.Argument(
+                name=ast_types.Name(value='op_name'),
+                value=ast_types.StringValue(value='in_collection'),
+            ),
+            ast_types.Argument(
+                name=ast_types.Name(value='value'),
+                value=ast_types.ListValue(
+                    values=[
+                        ast_types.StringValue(value=u'$' + input_filter_name),
+                    ],
+                ),
+            ),
+        ],
+    )
+
+
+def print_query_plan(query_plan_descriptor):
+    """Return a string describing query plan."""
+    query_plan_str = u''
+    plan_and_depth = _get_plan_and_depth_in_dfs_order(query_plan_descriptor.root_sub_query_plan)
+
+    for query_plan, depth in plan_and_depth:
+        line_separation = u'\n' + u' ' * 8 * depth
+        query_plan_str += line_separation
+
+        query_str = u'Execute in schema named "{}":\n'.format(query_plan.schema_id)
+        query_str += print_ast(query_plan.query_ast)
+        query_str = query_str.replace(u'\n', line_separation)
+        query_plan_str += query_str
+
+    query_plan_str += u'\n\n'
+    query_plan_str += u'Join together outputs as follows: '
+    query_plan_str += str(query_plan_descriptor.output_join_descriptors) + u'\n\n'
+    query_plan_str += u'Remove the following outputs at the end: '
+    query_plan_str += str(query_plan_descriptor.intermediate_output_names) + u'\n'
+
+    return query_plan_str
```

Comment:
nit: All this string concatenation is actually quite expensive since it uses Shlemiel the Painter's algorithm: https://www.joelonsoftware.com/2001/12/11/back-to-basics/

It would be better to assemble all the substrings we'd like to join together into a list, and then just call `''.join(query_plan_components)` with the list at the end.